proposals/shared-everything-threads/Overview.md

```diff
@@ -138,11 +124,11 @@ The abstract syntax of `tabletype` is extended:
 tabletype ::= share index_type limits reftype
 ```
 
-Similarly, the text format is extended:
+Similarly, the text format is extended. For consistency with the text format for
+shared memories, the optional `shared` keyword is placed after the limits.
 
 ```
-tabletype ::= it:index_type lim:limits rt:reftype => unshared it lim rt
-            | '(' 'shared' it:index_type lim:limits rt:reftype ')' => shared it lim rt
+tabletype ::= it:index_type lim:limits sh:share rt:reftype => sh it lim rt
```

Comment:
I really don't like this, it's even worse than putting it at the end. Why in the middle?

I'm afraid I really don't see what the placement has to do with the absence or presence of outer parens. Can't we just leave it as a prefix, like everywhere else?

Reply:
We have three options here:

 1. We could put `shared` first in table types and keep the current memory syntax. These would  be valid:
  ```
  (memory $m 1 1 shared)
  (table $t shared 1 1 anyref)
  ```
  Notably, `(memory $m shared 1 1)` would not be valid, so this is inconsistent.
  
 2. We could do the same but redefine the current memory syntax to be an abbreviation for the new, consistent syntax. These would be valid:
  ```
  (memory $m1 1 1 shared)
  (memory $m2 shared 1 1)
  (table $t shared 1 1 anyref)
  ```
  Note that there is still an inconsistency because tables would not have the abbreviation to put the `shared` after the limits.
  
 3. We could keep the existing memory syntax and choose the new table syntax to match it. These would be valid:
 ```
 (memory $m 1 1 shared)
 (table $t 1 1 shared anyref)
 ```
 This is consistent, even if it is odd that the `shared` does not go first.


Of these options, (3) seems the least bad to me. Which would you prefer?

Reply:
But why not `(table $t 1 1 anyref shared)` in the last case? Or vice versa, where would the custom page size go?

Personally, I'd prefer 2, because that's most logical and consistent. Perhaps even deprecating the previous not-yet-merged syntax?

Reply:
Sure, I don't have a preference between putting `shared` at the end or after the limits. Either is consistent with memories.

I wouldn't mind 2, so I'll add a commit switching to that solution. If we want to change the syntax for shared memories, we should check with the CG about the feasibility of that soon.

Reply:
To throw a very small wrench in things, my experience in `wasm-tools` shows me that we will have some weirdness due to duplication of `shared`. I implemented the shorthand form of references (e.g., `(shared anyref)`, meaning a reference of type `shared any` which @rossberg didn't like above); this meant that globals, e.g., would look like `(global (shared (shared anyref)) ...)`.

The reason I bring that up is that `(table $t 1 1 shared anyref)` above (and others like that) are invalid: if we still include shorthands, it actually must look like  `(table $t 1 1 shared (shared anyref))` and if we only allow `shared` in longhand from then it looks like `(table $t 1 1 shared (ref null (shared any))))`. Someone may suggest, "well, why doesn't the `table` shared-ness propagate down to its contained types?" but that seemed even weirder from a parsing perspective.

Reply:
`(table $t 1 1 shared (ref null (shared any))))` or `(table $t 1 1 (ref null (shared any)) shared)` seem fine to me. Is it just the repetition of `shared` that looks weird, or does it cause extra complexity in the parser somehow?

Reply:
Just the repetition. It looked even weirder in the shorthand form, but if we get rid of that then it doesn't look as confusing.

Reply:
> Someone may suggest, "well, why doesn't the table shared-ness propagate down to its contained types?" but that seemed even weirder from a parsing perspective.

Just to say, it wouldn't just be weird, it would be wrong. All 4 combinations of sharedness are representable in the binary format and the AST, hence they must be distinguishable in the text format, including combinations that are invalid — they are still needed for expressing corresponding negative tests.